tree/readspeed/src/ReadSpeed.cxx

```diff
@@ -0,0 +1,398 @@
+ByteData ReadSpeed::ReadTree(TFile *f, const std::string &treeName, const std::vector<std::string> &branchNames,
+                             EntryRange range)
+{
+   std::unique_ptr<TTree> t(f->Get<TTree>(treeName.c_str()));
+   if (t == nullptr)
+      throw std::runtime_error("Could not retrieve tree '" + treeName + "' from file '" + f->GetName() + '\'');
+
+   t->SetBranchStatus("*", 0);
+
+   std::vector<TBranch *> branches;
+   for (const auto &bName : branchNames) {
+      auto *b = t->GetBranch(bName.c_str());
+      if (b == nullptr)
+         throw std::runtime_error("Could not retrieve branch '" + bName + "' from tree '" + t->GetName() +
+                                  "' in file '" + t->GetCurrentFile()->GetName() + '\'');
+
+      b->SetStatus(1);
+      branches.push_back(b);
+   }
+
+   const auto nEntries = t->GetEntries();
+   if (range.fStart == -1ll)
+      range = EntryRange{0ll, nEntries};
+   else if (range.fEnd > nEntries)
+      throw std::runtime_error("Range end (" + std::to_string(range.fEnd) + ") is beyond the end of tree '" +
+                               t->GetName() + "' in file '" + t->GetCurrentFile()->GetName() + "' with " +
+                               std::to_string(nEntries) + " entries.");
+
+   ULong64_t bytesRead = 0;
+   const ULong64_t fileStartBytes = f->GetBytesRead();
+   for (auto e = range.fStart; e < range.fEnd; ++e)
+      for (const auto &b : branches)
+         bytesRead += b->GetEntry(e);
+
+   const ULong64_t fileBytesRead = f->GetBytesRead() - fileStartBytes;
+   return {bytesRead, fileBytesRead};
+}
+
+Result ReadSpeed::EvalThroughputST(const Data &d)
+{
+   auto treeIdx = 0;
+   ULong64_t uncompressedBytesRead = 0;
+   ULong64_t compressedBytesRead = 0;
+
+   TStopwatch sw;
+
+   std::vector<ReadSpeedRegex> regexes;
+   if (d.fUseRegex)
+      std::transform(d.fBranchNames.begin(), d.fBranchNames.end(), std::back_inserter(regexes), [](std::string text) {
+         return ReadSpeedRegex{text, std::regex(text)};
+      });
+
+   for (const auto &fileName : d.fFileNames) {
+      std::vector<std::string> branchNames;
+      if (d.fUseRegex)
+         branchNames = GetMatchingBranchNames(fileName, d.fTreeNames[treeIdx], regexes);
+      else
+         branchNames = d.fBranchNames;
+
+      TFile *f = TFile::Open(fileName.c_str(), "READ_WITHOUT_GLOBALREGISTRATION");
+      if (f == nullptr || f->IsZombie())
+         throw std::runtime_error("Could not open file '" + fileName + '\'');
+
+      sw.Start(kFALSE);
+
+      try {
+         const auto byteData = ReadTree(f, d.fTreeNames[treeIdx], branchNames);
+         uncompressedBytesRead += byteData.fUncompressedBytesRead;
+         compressedBytesRead += byteData.fCompressedBytesRead;
+      } catch (const std::runtime_error &err) {
+         if (f != nullptr && !f->IsZombie())
+            f->Close();
+         delete f;
+         throw;
+      }
```

Comment:
we can avoid having to write a `catch` clause by storing the result of `TFile::Open` in a `unique_ptr<TFile>` (which guarantees the `TFile`'s dtor is always called, even in case of an exception passing through -- and TFile closes itself on destruction)